device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/fileupload/FileUploadTask.java

```diff
@@ -199,9 +208,31 @@ public IotHubStatusCode sendNotification(FileUploadCompletionNotification fileUp
         ResponseMessage responseMessage = httpsTransportManager.sendFileUploadNotification(message);
         httpsTransportManager.close();
 
+        validateServiceStatusCode(responseMessage, "Failed to complete the file upload notification");
+
         return responseMessage.getStatus();
     }
 
+    private void validateServiceStatusCode(ResponseMessage responseMessage, String errorMessage) throws IOException
+    {
+        String responseMessagePayload = null;
+        if (responseMessage.getBytes() != null && responseMessage.getBytes().length > 0)
+        {
+            responseMessagePayload = new String(responseMessage.getBytes(), DEFAULT_IOTHUB_MESSAGE_CHARSET);
+        }
+
+        IotHubServiceException serviceException =
+            IotHubStatusCode.getConnectionStatusException(
+                responseMessage.getStatus(),
+                responseMessagePayload);
+
+        // serviceException is only not null if the provided status code was a non-successful status code like 400, 429, 500, etc.
+        if (serviceException != null)
+        {
+            throw new IOException(errorMessage, serviceException);
```

Comment:
These file upload methods only declare that they throw IOExceptions, so the closest thing we can do to throwing an IotHubServiceException is to nest it within an IOException.

Reply:
why not pull the specific error message out of the service exception returned, rather then it being generic text?

Reply:
That specific error message is embedded within the nested IoTHubServiceException, rather than the top level IOException